### DISCUSSION TITLE
Java version detected from gradle-wrapper is inconsistent

### DISCUSSION THREAD

**Discussion by gmariottiAtWolt** [renovatebot/renovate#18334] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.236.2

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

Looking at the logs, it seems that Renovate (as a Github Action) is inconsistent when deciding which Java version to use for a Gradle project. Looking at the source code, [this](https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/gradle-wrapper/utils.ts#L46-L58) should be the section that takes care of deciding the Java version to use. 
I don't have yet a reproducible project, but for a Gradle project using Java 17, the following `gradle-wrapper` configuration is detected (which is correct)

```json
"gradle-wrapper": [
           {
             "packageFile": "gradle/wrapper/gradle-wrapper.properties",
             "deps": [
               {
                 "depName": "gradle",
                 "currentValue": "7.5.1",
                 "replaceString": "https\\://services.gradle.org/distributions/gradle-7.5.1-bin.zip",
                 "datasource": "gradle-version",
                 "versioning": "gradle",
                 "depIndex": 0,
                 "updates": [],
                 "warnings": [],
                 "sourceUrl": "https://github.com/gradle/gradle",
                 "homepage": "https://gradle.org/",
                 "currentVersion": "7.5.1",
                 "fixedVersion": "7.5.1"
               }
             ]
           }
```

However, Renovate tries to update some dependencies with `installing v2 tool java v11.0.16+8` and others with `installing v2 tool java v8.0.345+1`

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by viceice** [comment, marked as answer]
do you see any errors using an older java version to update the gradle wrapper?

how should renovate know to use java v17 to update Gradle?

**Reply by gmariottiAtWolt** [reply]
I'm seeing errors with how the lockfile is updated but I can't point to the issue being Gradle or the Java version used. The only thing I noticed is this inconsistency, that during the same Renovate execution, it used Java 11 for certain dependencies and Java 8 for others but couldn't see from the logs what is making Renovate take this decision.

Regarding the JVM version to use, there is probably more than one way to figure it out from the Gradle setup but my suggestion would be to keep it simple and let the user override it with a an environment variable like `RENOVATE_JAVA_VERSION`. Would that be possible?

**Reply by viceice** [reply]
renovate uses the gradle wrapper major version, which ist statically mapped to a java major version.

checkout constraints config option for forcing specific tool versions.

you didn't provide any reproduction nor any relevant log lines, so i still say there's no bug but  a configuration issue on your repo.

**Reply by gmariottiAtWolt** [reply]
I have more than 4000 lines of debug logs with company related information, would it be possible for you to point me at what line I can look for that would help you understand why Renovate picked Java 11 and Java 8 over Java 17 (which is what I think it should pick based on https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/gradle-wrapper/utils.ts#L46-L58)?

**Reply by viceice** [reply]
the lines around the execution of Gradle? there should be the error logged. and before it should log the tool versions to install

**Reply by gmariottiAtWolt** [reply]
This is only what I was able to find (`***` is for redacting the repository and related stuff)

```
DEBUG: manager.getUpdatedPackageFiles() reuseExistingBranch=false (repository=***, branch=***)
DEBUG: Updating *** in build.gradle.kts (repository=***, branch=***)
DEBUG: gradle.updateArtifacts(build.gradle.kts) (repository=***, branch=***)
DEBUG: Updating found Gradle dependency lockfiles (repository=***, branch=***)
DEBUG: Setting CONTAINERBASE_CACHE_DIR to /tmp/renovate/cache/containerbase (repository=***, branch=***)
DEBUG: Using docker to execute (repository=***, branch=***)
       "image": "sidecar"
DEBUG: No concurrency limits (repository=***, branch=***)
       "host": "api.adoptium.net"
DEBUG: Resolved stable matching version (repository=***, branch=***)
       "toolName": "java",
       "constraint": "^11.0.0",
       "resolvedVersion": "11.0.16+8"
DEBUG: containerbaseDir is inside cacheDir (repository=***, branch=***)
DEBUG: No tag or tagConstraint specified (repository=***, branch=***)
       "image": "docker.io/renovate/sidecar"
DEBUG: Fetching Docker image: docker.io/renovate/sidecar (repository=***, branch=***)
DEBUG: Finished fetching Docker image docker.io/renovate/sidecar@sha256:d014556ad8450b0331b8a03f685fbe75b592dd52952d8271d0087b728d689fb2 (repository=***, branch=***)
DEBUG: Executing command (repository=***, branch=***)
       "command": "docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/tmp/renovate/repos/github/***\":\"/tmp/renovate/repos/github/***\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e GRADLE_OPTS -e ORG_GRADLE_PROJECT_artifactorySaasUsername -e ORG_GRADLE_PROJECT_artifactorySaasPassword -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/tmp/renovate/repos/github/***\" docker.io/renovate/sidecar bash -l -c \"install-tool java 11.0.16+8 && ./gradlew --console=plain -q properties\""
DEBUG: exec completed (repository=***, branch=***)
       "cmd": "docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/tmp/renovate/repos/github/***\":\"/tmp/renovate/repos/github/***\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e GRADLE_OPTS -e ORG_GRADLE_PROJECT_artifactorySaasUsername -e ORG_GRADLE_PROJECT_artifactorySaasPassword -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/tmp/renovate/repos/github/***\" docker.io/renovate/sidecar bash -l -c \"install-tool java 11.0.16+8 && ./gradlew --console=plain -q properties\"",
       "durationMs": 89056,
       "stdout": "installing v2 tool java v11.0.16+8\nlinking tool java v11.0.16+8\nInstalled v2 /usr/local/buildpack/tools/v2/java.sh in 3 seconds\n***REDACTED***\n",
       "stderr": "openjdk version \"11.0.16\" 2022-07-19\nOpenJDK Runtime Environment Temurin-11.0.16+8 (build 11.0.16+8)\nOpenJDK 64-Bit Server VM Temurin-11.0.16+8 (build 11.0.16+8, mixed mode)\n"
DEBUG: Setting CONTAINERBASE_CACHE_DIR to /tmp/renovate/cache/containerbase (repository=***, branch=***)
DEBUG: Using docker to execute (repository=***, branch=***)
       "image": "sidecar"
DEBUG: Resolved stable matching version (repository=***, branch=***)
       "toolName": "java",
       "constraint": "^11.0.0",
       "resolvedVersion": "11.0.16+8"
DEBUG: containerbaseDir is inside cacheDir (repository=***, branch=***)
DEBUG: No tag or tagConstraint specified (repository=***, branch=***)
       "image": "docker.io/renovate/sidecar"
DEBUG: Docker image is already prefetched: docker.io/renovate/sidecar@sha256:d014556ad8450b0331b8a03f685fbe75b592dd52952d8271d0087b728d689fb2 (repository=***, branch=***)
DEBUG: Executing command (repository=***, branch=***)
       "command": "docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/tmp/renovate/repos/github/***\":\"/tmp/renovate/repos/github/***\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e GRADLE_OPTS -e ORG_GRADLE_PROJECT_artifactorySaasUsername -e ORG_GRADLE_PROJECT_artifactorySaasPassword -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/tmp/renovate/repos/github/***\" docker.io/renovate/sidecar bash -l -c \"install-tool java 11.0.16+8 && ./gradlew --console=plain -q :dependencies --update-locks ***\""
DEBUG: exec completed (repository=***, branch=***)
       "cmd": "docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/tmp/renovate/repos/github/***\":\"/tmp/renovate/repos/github/***\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e GRADLE_OPTS -e ORG_GRADLE_PROJECT_artifactorySaasUsername -e ORG_GRADLE_PROJECT_artifactorySaasPassword -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/tmp/renovate/repos/github/***\" docker.io/renovate/sidecar bash -l -c \"install-tool java 11.0.16+8 && ./gradlew --console=plain -q :dependencies --update-locks ***\"",
       "durationMs": 661905,
       "stdout": "installing v2 tool java v11.0.16+8\nlinking tool java v11.0.16+8\nInstalled v2 /usr/local/buildpack/tools/v2/java.sh in 2 seconds\nskip cleanup, not a docker build: 97f57cc789b8\n\n
```

**Reply by viceice** [reply]
i don't see any error there. also.its not Gradle wrapper manager, it's the normal gradle manager with the new lockfile update feature.

**Reply by viceice** [reply]
ok, it seems here is a bug:
https://github.com/renovatebot/renovate/blob/66f35c954bc7ca70bfcdd0ced9ac30e11b8a56d0/lib/modules/manager/gradle/artifacts.ts#L111

it passes a wrong version to the java detection.


so your workaround it to configure constraints as told before.

**Reply by gmariottiAtWolt** [reply]
Thanks for finding it 🙏 so the following should be enough, right?

```json
"force": {
  "constraints": {
    "java": "> 17"
  }
}
```

Also, regarding the bug, should I open a separate issue?

**Reply by Churro** [reply]
The bug is addressed in PR https://github.com/renovatebot/renovate/pull/18346 👍

**Reply by gmariottiAtWolt** [reply]
Great, thank you 🙏

**Comment by viceice** [comment]
I've converted to discussion as you don't clearly show any bug. currently the gradle wrapper updates are working as expected.
